File: crates/cs/csharp/ZeroDDS.Cdr/src/Xcdr2Reader.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ZeroDDS Contributors
//
// XCDR2 Decoder.
// Spec: OMG XTypes 1.3 §7.4 + zerodds-xcdr2-csharp-1.0 §6/§7
// + zerodds-xcdr2-bindings-conformance-1.0 §6 (V-1..V-12).

using System;
using System.Buffers.Binary;
using System.Text;

namespace ZeroDDS.Cdr;

/// <summary>
/// XCDR2 decoder. Inverse of <see cref="Xcdr2Writer"/>.
///
/// Per XTypes 1.3 §7.4.1.5 the alignment rule is relative to the current
/// origin (initially 0; reset at the DHEADER body).
/// </summary>
public ref struct Xcdr2Reader
{
    private readonly ReadOnlySpan<byte> _buf;
    private readonly EndianMode _endian;
    private int _pos;
    private int _origin;

    /// <summary>
    /// XCDR2 maximum alignment (XTypes 1.3 §7.4.1.1.1): 8-byte primitives align
    /// to 4, never 8. Mirrors the zerodds-cdr core
    /// (`crates/cdr/src/buffer.rs`: `alignment.min(self.max_alignment)`).
    /// </summary>
    private const int Xcdr2MaxAlignment = 4;
    private const int Xcdr1MaxAlignment = 8;

    /// <summary>
    /// Effective alignment cap: 4 for XCDR2 (8-byte primitives align to 4), 8 for
    /// XCDR1 / classic CDR. Set from the representation at construction.
    /// </summary>
    private readonly int _maxAlignment;

    /// <summary>Constructor with default endianness (little-endian), XCDR2.</summary>
    public Xcdr2Reader(ReadOnlySpan<byte> bytes) : this(bytes, EndianMode.LittleEndian) { }

    /// <summary>Constructor with explicit endianness, XCDR2 representation.</summary>
    public Xcdr2Reader(ReadOnlySpan<byte> bytes, EndianMode endian)
        : this(bytes, endian, Xcdr2MaxAlignment) { }

    /// <summary>
    /// Constructor with explicit endianness and alignment cap. Pass
    /// <see cref="Xcdr1MaxAlignmentValue"/> (8) to read the XCDR1 / classic CDR
    /// wire (no DHEADER, PL_CDR1 for @mutable); the default 4 reads XCDR2.
    /// </summary>
    public Xcdr2Reader(ReadOnlySpan<byte> bytes, EndianMode endian, int maxAlignment)
    {
        _buf = bytes;
        _endian = endian;
        _pos = 0;
        _origin = 0;
        _maxAlignment = maxAlignment == Xcdr1MaxAlignment ? Xcdr1MaxAlignment : Xcdr2MaxAlignment;
    }

    /// <summary>XCDR1 alignment-cap value (8) for the 3-arg constructor.</summary>
    public const int Xcdr1MaxAlignmentValue = Xcdr1MaxAlignment;

    /// <summary>Active endianness.</summary>
    public EndianMode Endian => _endian;

    /// <summary>`true` when reading the XCDR1 / classic CDR wire (alignment cap 8,
    /// no DHEADER, PL_CDR1 for @mutable). Generated decoders branch on it.</summary>
    public bool IsXcdr1 => _maxAlignment == Xcdr1MaxAlignment;

    /// <summary>Current read byte position.</summary>
    public int Position => _pos;

    /// <summary>Number of bytes still available from `Position`.</summary>
    public int Remaining => _buf.Length - _pos;

    // ---------------------------------------------------------------------
    // Alignment + raw bytes
    // ---------------------------------------------------------------------

    /// <summary>Skips padding to the N-byte boundary relative to the origin.</summary>
    public void Align(int alignment)
    {
        if (alignment != 1 && alignment != 2 && alignment != 4 && alignment != 8)
        {
            throw new ArgumentOutOfRangeException(nameof(alignment),
                "alignment must be one of {1,2,4,8}");
        }
        // XCDR2 caps the effective alignment at 4 (§7.4.1.1.1); XCDR1 at 8.
        int effective = alignment < _maxAlignment ? alignment : _maxAlignment;
        int offset = _pos - _origin;
        int pad = (effective - (offset % effective)) % effective;
        if (_pos + pad > _buf.Length)
        {
            throw new XcdrException(
                $"alignment skip past end-of-buffer (pos={_pos}, pad={pad}, len={_buf.Length})");
        }
        _pos += pad;
    }

    /// <summary>Returns a slice of the next N bytes without endianness conversion.</summary>
    public ReadOnlySpan<byte> ReadBytes(int count)
    {
        if (count < 0) throw new ArgumentOutOfRangeException(nameof(count));
        if (_pos + count > _buf.Length)
        {
            throw new XcdrException(
                $"read past end-of-buffer (pos={_pos}, count={count}, len={_buf.Length})");
        }
        var slice = _buf.Slice(_pos, count);
        _pos += count;
        return slice;
    }

    /// <summary>
    /// Reads an IDL <c>fixed&lt;P,S&gt;</c>: (P+2)/2 packed-BCD octets back into a
    /// <see cref="decimal"/>. Inverse of <c>Xcdr2Writer.WriteFixedBcd</c>.
    /// </summary>
    public decimal ReadFixedBcd(int p, int s)
    {
        int n = (p + 2) / 2;
        var raw = ReadBytes(n);
        var chars = new System.Collections.Generic.List<char>();
        char sign = '+';
        for (int i = 0; i < n; i++)
        {
            int hi = (raw[i] >> 4) & 0x0F;
            int lo = raw[i] & 0x0F;
            chars.Add((char)('0' + (hi % 10)));
            if (i == n - 1) sign = lo == 0x0D ? '-' : '+';
            else chars.Add((char)('0' + (lo % 10)));
        }
        while (chars.Count > s + 1 && chars[0] == '0') chars.RemoveAt(0);
        var sb = new System.Text.StringBuilder();
        if (sign == '-') sb.Append('-');
        if (s > 0)
        {
            int dot = Math.Max(chars.Count - s, 0);
            for (int i = 0; i < chars.Count; i++)
            {
                if (i == dot) sb.Append('.');
                sb.Append(chars[i]);
            }
        }
        else
        {
            sb.Append(chars.ToArray());
        }
        return decimal.Parse(sb.ToString(), System.Globalization.CultureInfo.InvariantCulture);
    }

    // ---------------------------------------------------------------------
    // Primitives
    // ---------------------------------------------------------------------

    /// <summary>IDL `boolean` -> 1 byte.</summary>
    public bool ReadBool()
    {
        var b = ReadByte();
        if (b != 0 && b != 1)
        {
            throw new XcdrException($"invalid boolean encoding 0x{b:X2}");
        }
        return b != 0;
    }

    /// <summary>IDL `octet` / `char` -> 1 byte.</summary>
    public byte ReadByte()
    {
        if (_pos >= _buf.Length)
        {
            throw new XcdrException("read past end-of-buffer");
        }
        return _buf[_pos++];
    }

    /// <summary>IDL `octet` (alias of ReadByte for symmetry with the writer).</summary>
    public byte ReadOctet() => ReadByte();

    /// <summary>IDL `short` -> 2 bytes, Align(2).</summary>
    public short ReadInt16()
    {
        Align(2);
        var s = ReadBytes(2);
        return _endian == EndianMode.LittleEndian
            ? BinaryPrimitives.ReadInt16LittleEndian(s)
            : BinaryPrimitives.ReadInt16BigEndian(s);
    }

    /// <summary>IDL `unsigned short` -> 2 bytes, Align(2).</summary>
    public ushort ReadUInt16()
    {
        Align(2);
        var s = ReadBytes(2);
        return _endian == EndianMode.LittleEndian
            ? BinaryPrimitives.ReadUInt16LittleEndian(s)
            : BinaryPrimitives.ReadUInt16BigEndian(s);
    }

    /// <summary>IDL `long` -> 4 bytes, Align(4).</summary>
    public int ReadInt32()
    {
        Align(4);
        var s = ReadBytes(4);
        return _endian == EndianMode.LittleEndian
            ? BinaryPrimitives.ReadInt32LittleEndian(s)
            : BinaryPrimitives.ReadInt32BigEndian(s);
    }

    /// <summary>IDL `unsigned long` -> 4 bytes, Align(4).</summary>
    public uint ReadUInt32()
    {
        Align(4);
        var s = ReadBytes(4);
        return _endian == EndianMode.LittleEndian
            ? BinaryPrimitives.ReadUInt32LittleEndian(s)
            : BinaryPrimitives.ReadUInt32BigEndian(s);
    }

    /// <summary>IDL `long long` -> 8 bytes, Align(8).</summary>
    public long ReadInt64()
    {
        Align(8);
        var s = ReadBytes(8);
        return _endian == EndianMode.LittleEndian
            ? BinaryPrimitives.ReadInt64LittleEndian(s)
            : BinaryPrimitives.ReadInt64BigEndian(s);
    }

    /// <summary>IDL `unsigned long long` -> 8 bytes, Align(8).</summary>
    public ulong ReadUInt64()
    {
        Align(8);
        var s = ReadBytes(8);
        return _endian == EndianMode.LittleEndian
            ? BinaryPrimitives.ReadUInt64LittleEndian(s)
            : BinaryPrimitives.ReadUInt64BigEndian(s);
    }

    /// <summary>IDL `float` -> 4 bytes IEEE-754, Align(4).</summary>
    public float ReadFloat32()
    {
        Align(4);
        var s = ReadBytes(4);
        return _endian == EndianMode.LittleEndian
            ? BinaryPrimitives.ReadSingleLittleEndian(s)
            : BinaryPrimitives.ReadSingleBigEndian(s);
    }

    /// <summary>IDL `double` -> 8 bytes IEEE-754, Align(8).</summary>
    public double ReadFloat64()
    {
        Align(8);
        var s = ReadBytes(8);
        return _endian == EndianMode.LittleEndian
            ? BinaryPrimitives.ReadDoubleLittleEndian(s)
            : BinaryPrimitives.ReadDoubleBigEndian(s);
    }

    /// <summary>IDL `wchar` -> 2-byte UTF-16 code unit, Align(2).</summary>
    public char ReadWChar() => (char)ReadUInt16();

    // ---------------------------------------------------------------------
    // String / wstring
    // ---------------------------------------------------------------------

    /// <summary>
    /// IDL `string` -> uint32 length-incl-NUL + UTF-8 bytes + NUL.
    /// Throws if the terminating NUL is missing.
    /// </summary>
    public string ReadString()
    {
        uint len = ReadUInt32();
        if (len == 0)
        {
            throw new XcdrException("string length must be >= 1 (NUL terminator required)");
        }
        var bytes = ReadBytes((int)len);
        // The last byte MUST be NUL.
        if (bytes[bytes.Length - 1] != 0)
        {
            throw new XcdrException("string is not NUL-terminated");
        }
        return Encoding.UTF8.GetString(bytes.Slice(0, bytes.Length - 1));
    }

    /// <summary>
    /// IDL `wstring` -> uint32 length (code units, no NUL) + UTF-16-LE code units.
    /// </summary>
    public string ReadWString()
    {
        uint len = ReadUInt32();
        var sb = new StringBuilder((int)len);
        for (uint i = 0; i < len; i++)
        {
            sb.Append((char)ReadUInt16());
        }
        return sb.ToString();
    }

    /// <summary>Reads the sequence counter (uint32, Align(4)).</summary>
    public int ReadSequenceLength()
    {
        uint len = ReadUInt32();
        if (len > int.MaxValue)
        {
            throw new XcdrException($"sequence length overflow: {len}");
        }
        return (int)len;
    }

    // ---------------------------------------------------------------------
    // DHEADER + EMHEADER
    // ---------------------------------------------------------------------

    /// <summary>
    /// Reads the 4-byte DHEADER (object-size in bytes), sets a new
    /// alignment origin at the position right after the header, and
    /// returns a token for restore + bound check.
    /// </summary>
    public DHeaderReadScope BeginDHeader()
    {
        // XCDR1 / classic CDR has no DHEADER — an @appendable/@final aggregate
        // continues in the same stream. Return a frame-less scope so the matching
        // EndDHeader is a no-op and member reads run over the current stream.
        if (_maxAlignment != Xcdr2MaxAlignment)
        {
            return new DHeaderReadScope(_pos, _buf.Length, _origin, hasFrame: false);
        }
        Align(4);
        uint size = ReadUInt32();
        int previousOrigin = _origin;
        _origin = _pos;
        if (_pos + size > _buf.Length)
        {
            throw new XcdrException(
                $"DHEADER size {size} exceeds buffer (pos={_pos}, len={_buf.Length})");
        }
        return new DHeaderReadScope(_pos, _origin + (int)size, previousOrigin);
    }

    /// <summary>
    /// Closes the DHEADER scope: skips any remaining trailing padding
    /// up to the object-size end and restores the previous origin.
    /// </summary>
    public void EndDHeader(DHeaderReadScope scope)
    {
        if (!scope.HasFrame)
        {
            return; // XCDR1: no frame to close.
        }
        if (_pos > scope.BodyEnd)
        {
            throw new XcdrException(
                $"read past DHEADER body-end (pos={_pos}, body-end={scope.BodyEnd})");
        }
        _pos = scope.BodyEnd;
        _origin = scope.PreviousOrigin;
    }

    /// <summary>
    /// Begins one PL_CDR1 (@mutable XCDR1) member: a 4-byte aligned header
    /// [u16 PID][u16 length], then shifts the alignment origin to the body start
    /// (PL_CDR1 bodies align member-relative, like the cdr-core fresh-reader
    /// path) so the member's own decoder reads inline from this reader. Returns
    /// <c>false</c> at the PID_LIST_END sentinel. PID_EXTENDED (length 8) carries
    /// a 32-bit member id + 32-bit body length. Pair with
    /// <see cref="EndPlCdr1Member"/>. Mirrors cdr-core `xcdr1::read_pl_cdr1_member`.
    /// </summary>
    public bool BeginPlCdr1Member(out uint memberId, out DHeaderReadScope scope)
    {
        const ushort PidListEnd = 0x3F02;
        const ushort PidExtended = 0x3F01;
        Align(4);
        if (_pos + 4 > _buf.Length)
        {
            memberId = 0;
            scope = default;
            return false;
        }
        ushort pid = ReadUInt16();
        ushort lenU16 = ReadUInt16();
        if (pid == PidListEnd)
        {
            memberId = 0;
            scope = default;
            return false;
        }
        int bodyLen;
        if (pid == PidExtended)
        {
            memberId = ReadUInt32();
            bodyLen = (int)ReadUInt32();
        }
        else
        {
            memberId = pid;
            bodyLen = lenU16;
        }
        if (_pos + bodyLen > _buf.Length)
        {
            throw new XcdrException(
                $"PL_CDR1 member body {bodyLen} exceeds buffer (pos={_pos}, len={_buf.Length})");
        }
        int prevOrigin = _origin;
        _origin = _pos; // member-relative alignment
        scope = new DHeaderReadScope(_pos, _pos + bodyLen, prevOrigin);
        return true;
    }

    /// <summary>Closes a <see cref="BeginPlCdr1Member"/> scope: positions at the
    /// body end, skips the trailing 4-byte pad, and restores the origin.</summary>
    public void EndPlCdr1Member(DHeaderReadScope scope)
    {
        _pos = scope.BodyEnd;
        _origin = scope.PreviousOrigin;
        int bodyLen = scope.BodyEnd - scope.BodyStart;
        int pad = (4 - (bodyLen % 4)) % 4;
        for (int i = 0; i < pad && _pos < _buf.Length; i++)
        {
            _pos++;
        }
    }

    /// <summary>`true` when the current position has reached the body-end of the DHEADER scope.</summary>
    public bool DHeaderDone(DHeaderReadScope scope) => _pos >= scope.BodyEnd;

    /// <summary>
    /// Reads a 4-byte EMHEADER (ambient stream-endian per XTypes 1.3
    /// §7.4.3.4.5) and returns (memberId, lc, must_understand).
    /// </summary>
    public (uint MemberId, int Lc, bool MustUnderstand) ReadEmHeader()
    {
        uint header = ReadUInt32();
        bool mu = (header & 0x80000000u) != 0;
        int lc = (int)((header >> 28) & 0x7u);
        uint id = header & 0x0FFFFFFFu;
        return (id, lc, mu);
    }
}

/// <summary>
/// Token for a DHEADER scope while reading. Holds the absolute body-end
/// and the previous alignment origin.
/// </summary>
public readonly struct DHeaderReadScope
{
    /// <summary>Absolute buffer position of the body start (right after the 4-byte header).</summary>
    public int BodyStart { get; }

    /// <summary>Absolute buffer position right after the last body byte.</summary>
    public int BodyEnd { get; }

    /// <summary>Origin before the `BeginDHeader` call, for restore in `EndDHeader`.</summary>
    public int PreviousOrigin { get; }

    /// <summary>`false` for an XCDR1 reader, where aggregates carry no DHEADER —
    /// `BeginDHeader`/`EndDHeader` are then no-ops over the same stream.</summary>
    public bool HasFrame { get; }

    internal DHeaderReadScope(int bodyStart, int bodyEnd, int previousOrigin, bool hasFrame = true)
    {
        BodyStart = bodyStart;
        BodyEnd = bodyEnd;
        PreviousOrigin = previousOrigin;
        HasFrame = hasFrame;
    }
}

File: crates/java-omgdds/java/src/main/java/org/zerodds/cdr/Xcdr2Reader.java
// SPDX-License-Identifier: Apache-2.0
package org.zerodds.cdr;

import java.nio.charset.StandardCharsets;

/**
 * XCDR2 decoder for language bindings.
 *
 * <p>Inverse of {@link Xcdr2Writer}. Bounds checks on every read;
 * Stream-Underflow wirft {@link XcdrException}.
 */
public final class Xcdr2Reader {

    /** XCDR2 (4) / XCDR1 (8) maximum alignment (XTypes 1.3 §7.4.1.1.1). */
    private static final int XCDR2_MAX_ALIGN = 4;
    private static final int XCDR1_MAX_ALIGN = 8;

    private final byte[] buf;
    // Non-final: PL_CDR1 (@mutable XCDR1) shifts the alignment origin to each
    // member's body start so member bodies align member-relative, then restores.
    private int origin;
    private final int limit;
    private int pos;
    private final EndianMode endian;
    private final int maxAlign;

    /** Reads from {@code buf[0..buf.length]} with default LE. */
    public Xcdr2Reader(byte[] buf) {
        this(buf, 0, buf.length, EndianMode.LITTLE_ENDIAN);
    }

    /** Reads with the chosen endianness. */
    public Xcdr2Reader(byte[] buf, EndianMode endian) {
        this(buf, 0, buf.length, endian);
    }

    /** Reads from a subrange (XCDR2). Alignment is measured relative to {@code offset}. */
    public Xcdr2Reader(byte[] buf, int offset, int length, EndianMode endian) {
        this(buf, offset, length, endian, XCDR2_MAX_ALIGN);
    }

    /**
     * Reads from a subrange with an explicit alignment cap. Pass
     * {@link #XCDR1_MAX_ALIGN_VALUE} (8) to decode the XCDR1 / classic-CDR wire
     * (no DHEADER on aggregates, PL_CDR1 for {@code @mutable}); the default 4
     * reads XCDR2.
     */
    public Xcdr2Reader(byte[] buf, int offset, int length, EndianMode endian, int maxAlign) {
        if (buf == null) {
            throw new XcdrException("buffer is null");
        }
        if (offset < 0 || length < 0 || offset + length > buf.length) {
            throw new XcdrException(
                    "invalid sub-range: off=" + offset + " len=" + length + " buf.len=" + buf.length);
        }
        this.buf = buf;
        this.origin = offset;
        this.limit = offset + length;
        this.pos = offset;
        this.endian = endian;
        this.maxAlign = (maxAlign == XCDR1_MAX_ALIGN) ? XCDR1_MAX_ALIGN : XCDR2_MAX_ALIGN;
    }

    /** XCDR1 alignment-cap value (8) for the 5-arg constructor. */
    public static final int XCDR1_MAX_ALIGN_VALUE = XCDR1_MAX_ALIGN;

    /** {@code true} when reading the XCDR1 / classic CDR wire. */
    public boolean isXcdr1() {
        return maxAlign == XCDR1_MAX_ALIGN;
    }

    /** Current read position relative to origin. */
    public int position() {
        return pos - origin;
    }

    /** Verbleibende Bytes. */
    public int remaining() {
        return limit - pos;
    }

    // ------------------------------------------------------------------
    // Primitive Reader
    // ------------------------------------------------------------------

    public boolean readBoolean() {
        ensure(1);
        return buf[pos++] != 0;
    }

    public byte readOctet() {
        ensure(1);
        return buf[pos++];
    }

    public int readUInt8() {
        ensure(1);
        return buf[pos++] & 0xFF;
    }

    public char readChar() {
        return (char) readUInt8();
    }

    public char readWChar() {
        align(2);
        ensure(2);
        return (char) readShortRaw();
    }

    public short readInt16() {
        align(2);
        ensure(2);
        return readShortRaw();
    }

    public int readUInt16() {
        return readInt16() & 0xFFFF;
    }

    public int readInt32() {
        align(4);
        ensure(4);
        return readIntRaw();
    }

    public long readUInt32() {
        return readInt32() & 0xFFFF_FFFFL;
    }

    public long readInt64() {
        // 8-byte primitive: align to its natural 8; the central cap reduces it
        // to 4 for XCDR2 (Bug XW) and keeps 8 for XCDR1 / classic CDR.
        align(8);
        ensure(8);
        return readLongRaw();
    }

    public long readUInt64() {
        return readInt64();
    }

    public float readFloat32() {
        align(4);
        ensure(4);
        return Float.intBitsToFloat(readIntRaw());
    }

    public double readFloat64() {
        // 8-byte primitive: align(8); the central cap reduces it to 4 for XCDR2.
        align(8);
        ensure(8);
        return Double.longBitsToDouble(readLongRaw());
    }

    /** Reads a string per XTypes §7.4.4.6. */
    public String readString() {
        long total = readUInt32();
        if (total > Integer.MAX_VALUE) {
            throw new XcdrException("string length exceeds int range: " + total);
        }
        int len = (int) total;
        ensure(len);
        // Strip trailing NUL.
        int textLen = len;
        if (textLen > 0 && buf[pos + textLen - 1] == 0) {
            textLen--;
        }
        String s = new String(buf, pos, textLen, StandardCharsets.UTF_8);
        pos += len;
        return s;
    }

    /** Reads a WString (UTF-16-LE). */
    public String readWString() {
        long count = readUInt32();
        if (count > Integer.MAX_VALUE / 2) {
            throw new XcdrException("wstring length exceeds int range: " + count);
        }
        int n = (int) count;
        ensure(n * 2);
        char[] chars = new char[n];
        for (int i = 0; i < n; i++) {
            chars[i] = (char) readShortRaw();
        }
        return new String(chars);
    }

    /** Reads the sequence count (uint32). */
    public int readSequenceCount() {
        long count = readUInt32();
        if (count > Integer.MAX_VALUE) {
            throw new XcdrException("sequence count exceeds int range: " + count);
        }
        return (int) count;
    }

    /** Reads {@code len} raw bytes (no alignment). */
    public byte[] readBytes(int len) {
        if (len < 0) {
            throw new XcdrException("readBytes negative len: " + len);
        }
        ensure(len);
        byte[] out = new byte[len];
        System.arraycopy(buf, pos, out, 0, len);
        pos += len;
        return out;
    }

    /**
     * Reads an IDL {@code fixed<P,S>}: (P+2)/2 packed-BCD octets back into a
     * {@link java.math.BigDecimal}. Inverse of {@code Xcdr2Writer.writeFixedBcd}.
     */
    public java.math.BigDecimal readFixedBcd(int p, int s) {
        int n = (p + 2) / 2;
        byte[] raw = readBytes(n);
        StringBuilder chars = new StringBuilder();
        char sign = '+';
        for (int i = 0; i < n; i++) {
            int hi = (raw[i] >> 4) & 0x0F;
            int lo = raw[i] & 0x0F;
            chars.append((char) ('0' + (hi % 10)));
            if (i == n - 1) {
                sign = lo == 0x0D ? '-' : '+';
            } else {
                chars.append((char) ('0' + (lo % 10)));
            }
        }
        while (chars.length() > s + 1 && chars.charAt(0) == '0') chars.deleteCharAt(0);
        StringBuilder outStr = new StringBuilder();
        if (sign == '-') outStr.append('-');
        if (s > 0) {
            int dot = Math.max(chars.length() - s, 0);
            for (int i = 0; i < chars.length(); i++) {
                if (i == dot) outStr.append('.');
                outStr.append(chars.charAt(i));
            }
        } else {
            outStr.append(chars);
        }
        return new java.math.BigDecimal(outStr.toString());
    }

    // ------------------------------------------------------------------
    // Extensibility-Frames
    // ------------------------------------------------------------------

    /** Reads the DHEADER and returns the body size; the caller sets its own limit. */
    public int readDHeader() {
        // XCDR1 / classic CDR has no DHEADER — return 0 so the generated
        // `__endPos = position() + 0` equals the current position: the trailing
        // `while (position() < __endPos) skip` loop then does nothing, and an
        // @appendable/@final aggregate (top-level or nested) reads its members
        // inline without consuming a phantom DHEADER. (@mutable branches to the
        // PL_CDR1 reader in the generated decode.)
        if (isXcdr1()) {
            return 0;
        }
        long size = readUInt32();
        if (size > (long) (limit - pos)) {
            throw new XcdrException(
                    "DHEADER size " + size + " exceeds remaining " + (limit - pos));
        }
        return (int) size;
    }

    /** EMHEADER-Decoded Form. */
    public static final class EmHeader {
        public final int memberId;
        public final int lc;
        public final boolean mustUnderstand;

        EmHeader(int memberId, int lc, boolean mustUnderstand) {
            this.memberId = memberId;
            this.lc = lc;
            this.mustUnderstand = mustUnderstand;
        }
    }

    /** Reads the EMHEADER (XTypes §7.4.3.4.5). */
    public EmHeader readEmHeader() {
        align(4);
        ensure(4);
        int header = readIntRaw();
        boolean mu = (header & 0x8000_0000) != 0;
        int lc = (header >>> 28) & 0x7;
        int id = header & 0x0FFF_FFFF;
        return new EmHeader(id, lc, mu);
    }

    public int readNextInt() {
        align(4);
        ensure(4);
        return readIntRaw();
    }

    /** A PL_CDR1 (@mutable XCDR1) member: its id, the absolute end of its body,
     * and the alignment origin to restore when the member is closed. */
    public static final class PlCdr1Member {
        public final int memberId;
        public final int bodyEnd; // absolute buffer position right after the body
        final int prevOrigin;

        PlCdr1Member(int memberId, int bodyEnd, int prevOrigin) {
            this.memberId = memberId;
            this.bodyEnd = bodyEnd;
            this.prevOrigin = prevOrigin;
        }
    }

    /**
     * Begins one PL_CDR1 member: a 4-byte aligned {@code [u16 PID][u16 length]}
     * header; the member body then follows inline. Returns {@code null} at the
     * PID_LIST_END sentinel. PID_EXTENDED (length 8) carries a 32-bit member id +
     * 32-bit body length. Mirrors cdr-core {@code xcdr1::read_pl_cdr1_member}.
     */
    public PlCdr1Member beginPlCdr1Member() {
        final int PID_LIST_END = 0x3F02;
        final int PID_EXTENDED = 0x3F01;
        align(4);
        if (pos + 4 > limit) {
            return null;
        }
        int pid = readUInt16();
        int lenU16 = readUInt16();
        if (pid == PID_LIST_END) {
            return null;
        }
        int memberId;
        int bodyLen;
        if (pid == PID_EXTENDED) {
            memberId = (int) readUInt32();
            bodyLen = (int) readUInt32();
        } else {
            memberId = pid;
            bodyLen = lenU16;
        }
        if (pos + bodyLen > limit) {
            throw new XcdrException("PL_CDR1 member body " + bodyLen + " exceeds remaining " + (limit - pos));
        }
        // Member-relative alignment: shift the origin to the body start so an
        // 8-byte member inside aligns relative to the member, like the cdr-core
        // fresh-reader path. Restored by endPlCdr1Member.
        int prevOrigin = origin;
        origin = pos;
        return new PlCdr1Member(memberId, pos + bodyLen, prevOrigin);
    }

    /** Closes a {@link #beginPlCdr1Member}: restores the origin, positions at the
     * body end, and skips the 4-byte (stream-relative) pad. */
    public void endPlCdr1Member(PlCdr1Member m) {
        origin = m.prevOrigin;
        pos = m.bodyEnd;
        int rel = (pos - origin) & 3;
        int pad = (4 - rel) & 3;
        for (int i = 0; i < pad && pos < limit; i++) {
            pos++;
        }
    }

    public boolean readPresenceFlag() {
        return readBoolean();
    }

    // ------------------------------------------------------------------
    // Alignment
    // ------------------------------------------------------------------

    public void align(int boundary) {
        // Cap at the representation's max alignment (XCDR2 = 4, XCDR1 = 8).
        if (boundary > maxAlign) {
            boundary = maxAlign;
        }
        int rem = (pos - origin) % boundary;
        if (rem != 0) {
            int padBytes = boundary - rem;
            ensure(padBytes);
            pos += padBytes;
        }
    }

    /** Skip {@code n} bytes (no alignment). */
    public void skip(int n) {
        if (n < 0) {
            throw new XcdrException("skip negative: " + n);
        }
        ensure(n);
        pos += n;
    }

    private void ensure(int n) {
        if (pos + n > limit) {
            throw new XcdrException(
                    "underflow: need " + n + ", have " + (limit - pos) + " (pos=" + (pos - origin) + ")");
        }
    }

    private short readShortRaw() {
        if (endian == EndianMode.LITTLE_ENDIAN) {
            int b0 = buf[pos++] & 0xFF;
            int b1 = buf[pos++] & 0xFF;
            return (short) ((b1 << 8) | b0);
        } else {
            int b0 = buf[pos++] & 0xFF;
            int b1 = buf[pos++] & 0xFF;
            return (short) ((b0 << 8) | b1);
        }
    }

    private int readIntRaw() {
        if (endian == EndianMode.LITTLE_ENDIAN) {
            int b0 = buf[pos++] & 0xFF;
            int b1 = buf[pos++] & 0xFF;
            int b2 = buf[pos++] & 0xFF;
            int b3 = buf[pos++] & 0xFF;
            return (b3 << 24) | (b2 << 16) | (b1 << 8) | b0;
        } else {
            int b0 = buf[pos++] & 0xFF;
            int b1 = buf[pos++] & 0xFF;
            int b2 = buf[pos++] & 0xFF;
            int b3 = buf[pos++] & 0xFF;
            return (b0 << 24) | (b1 << 16) | (b2 << 8) | b3;
        }
    }

    private long readLongRaw() {
        if (endian == EndianMode.LITTLE_ENDIAN) {
            long v = 0;
            for (int i = 0; i < 8; i++) {
                v |= ((long) (buf[pos++] & 0xFF)) << (i * 8);
            }
            return v;
        } else {
            long v = 0;
            for (int i = 7; i >= 0; i--) {
                v |= ((long) (buf[pos++] & 0xFF)) << (i * 8);
            }
            return v;
        }
    }
}
